rfcs/20240808-graph-api-int-compression-for-sdpa/README.md

```diff
@@ -0,0 +1,307 @@
+      inputs should be `1d` tensors.
+   2. For `per_group` quantization, all dimensions should match the input,
+      excepts for the dimension where grouped quantization applies, which should
+      be `src_dim / group_size`.
+
+
+For example, if the shape of Key tensor is `4x8x4`, and the group size is `4`,
+which means each scaling factor will be adopted for 4 times. One possible scale
+would be:
+
+![alt text](img/grouped_scale.png)
+
+The usage of the new attribute will be like:
+
+```cpp
+using namespace dnnl::graph;
+
+const dim K = 10, N = 20;
+const size_t G = N / 2;
+
+dims src_dims = {K, N};
+dims scale_dims = {K, N/G};
+
+// In real case, users do not have to provide the concrete shapes of logical
+// tensors until compilation stage. 
+logical_tensor src_desc {0, data_type::u4, src_dims, layout_type::strided};
+logical_tensor scale_desc {
+        1, data_type::f32, scale_dims, layout_type::strided};
+logical_tensor dst_desc {2, data_type::f32, src_dims, layout_type::strided};
+
+op deq = op(0, op::kind::DynamicDequantize, {src_desc, scale_desc},
+        {dst_desc}, "deq");
+// Specify the quantization type as per_group quantization.
+deq.set_attr<std::string>(op::attr::qtype, "per_group");
+// Group size indicates the the size of each group.
+deq.set_attr<int64_t>(op::attr::group_size, G);
+
+// Create graph and add the op to the graph
+graph g(dnnl::engine::kind::gpu);
+g.add_op(deq);
+
+// ...
+```
+
+
+#### Alternative options
+
+Similar to the discussion of int8 and fp8 data types, an alternative solution
+may be supporting int4 data types directly in computation operations like MatMul
+and Convolution. But it will bloat and complicate the opset and op schema of
+oneDNN Graph. Hence it's not considered here.
+
+In addition, currrently oneDNN Graph does not support the direct int8
+quantization and dequantization between int8 and f16/bf16 data types. To align
+with current integration flow and maintain the consistency of library design,
+quantization and dequantization between int4 and f16/bf16 data types will also
+be implemented with a combination of quantization ops and `TypeCast`. An
+alternative solution may be to extend quantization ops to support data type
+conversion directly between int4 and f16/bf16 data types. But it will
+unnecessarily complicates the fusion patterns in oneDNN Graph. For example,
+an int4 quantization will be expressed in two different ways: 
+
+1. With a single `Quantize(f16 -> int4)`
+2. With two operations `TypeCast(f16 -> f32) -> Quantize(f32 -> int4)`
+
+It will require extensions for all patterns related to mixed data type
+quantization. Hence the option is not considered here.
+
+### Proposal 3: Extend SDPA pattern with compressed K/V
+
+#### Pattern definition
+
+With the reference to int4 GQA pattern in
+[PyTorch](https://pytorch.org/blog/int4-decoding/), the proposed fusion pattern
+for SDPA with compressed K/V in oneDNN Graph API is defined as:
+
+![alt text](img/sdpa_pattern.png)
```

Comment:
The picture illustrates scales in f32 and zero-points in s32. To provide clarity, it should specify different data types for scales as bf16/f16/f32 and different types for zero-points: u4/s4/u8/s8/s32. Please don't expect users to upconvert their scales to call Graph API.

Reply:
Thanks for reminding. The picture is updated.